### PR TITLE
feat: Added warmup each cycle feature in CyclicalScheduler

### PR DESCRIPTION
**Pull request by sihyeong671**
- add _get_param method in CyclicalScheduler
- add warmup_duration variable in CyclicalScheduler

issue #3036

Description:
this feature is following https://github.com/katsura-jp/pytorch-cosine-annealing-with-warmup

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/handlers/param_scheduler.py [hunk 1]
```diff
@@ -313,6 +315,18 @@ def __init__(
         self.cycle = 0
         self.start_value_mult = start_value_mult
         self.end_value_mult = end_value_mult
+        self.warmup_each_cycle = warmup_each_cycle
+        if not self.warmup_each_cycle:
+            if warmup_duration is not None:
+                warnings.warn(
+                    f"warmup_each_cycle=False but your warmup_duration is {warmup_duration}. "
+                    f"so warmup_duration will be set to 0. "
+                    f"If you want to use warmup each cycle, pleas set warmup_each_cycle=True"
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:324]:
```suggestion
                    "so warmup_duration will be set to 0. "
                    "If you want to use warmup each cycle, pleas set warmup_each_cycle=True"
```

File: ignite/handlers/param_scheduler.py [hunk 2]
```diff
@@ -313,6 +315,18 @@ def __init__(
         self.cycle = 0
         self.start_value_mult = start_value_mult
         self.end_value_mult = end_value_mult
+        self.warmup_each_cycle = warmup_each_cycle
+        if not self.warmup_each_cycle:
+            if warmup_duration is not None:
+                warnings.warn(
+                    f"warmup_each_cycle=False but your warmup_duration is {warmup_duration}. "
+                    "so warmup_duration will be set to 0. "
+                    "If you want to use warmup each cycle, pleas set warmup_each_cycle=True"
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:324]:
```suggestion
                    "If you want to use warmup each cycle, please set warmup_each_cycle=True"
```

File: ignite/handlers/param_scheduler.py
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:1]: Please update `CyclicalScheduler`'s docstring and add this to its end:
```

.. versionchanged:: 0.4.13
        Added warmup to the scheduler using ``warmup_each_cycle`` and ``warmup_duration``.
```

File: ignite/handlers/param_scheduler.py [hunk 3]
```diff
@@ -313,6 +315,18 @@ def __init__(
         self.cycle = 0
         self.start_value_mult = start_value_mult
         self.end_value_mult = end_value_mult
+        self.warmup_each_cycle = warmup_each_cycle
+        if not self.warmup_each_cycle:
+            if warmup_duration is not None:
+                warnings.warn(
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:321]: We should add proper tests to cover the new changes. For this uncovered line and also line 367, you could add this test to `tests/ignite/handlers/test_param_scheduler.py`:
```python
def test_cyclical_scheduler_asserts():
    tensor = torch.zeros([1], requires_grad=True)
    optimizer = torch.optim.SGD([tensor], lr=0)
    with pytest.warns(match="warmup_each_cycle=False but your warmup_duration"):
        scheduler = CyclicalScheduler(optimizer, "lr", 0., 1., 10, warmup_each_cycle=False, warmup_duration=2)

    with pytest.raises(NotImplementedError):
       scheduler._get_cycle_param()
```

File: ignite/handlers/param_scheduler.py [hunk 4]
```diff
@@ -536,9 +565,9 @@ def print_lr():
     .. versionadded:: 0.4.5
     """
 
-    def get_param(self) -> float:
+    def _get_cycle_param(self):
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:568]:
```suggestion
    def _get_cycle_param(self) -> float:
```

File: ignite/handlers/param_scheduler.py [hunk 5]
```diff
@@ -432,7 +461,7 @@ def print_lr():
...
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:464]:
```suggestion
    def _get_cycle_param(self) -> float:
```

File: ignite/handlers/param_scheduler.py [hunk 6]
```diff
@@ -536,9 +565,9 @@ def print_lr():
...
         """Method to get current optimizer's parameter value"""
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:569]: Something like
```suggestion
        """Method to get the cycle's current parameter value"""
```
?

File: ignite/handlers/param_scheduler.py [hunk 7]
```diff
@@ -325,18 +339,33 @@ def __init__(
...
 
+    def get_param(self) -> float:
+        """Method to get current optimizer's parameter value"""
+        if self.warmup_each_cycle and self.event_index < self.warmup_duration:
+            return self.end_value + (self.start_value - self.end_value) * self.event_index / self.warmup_duration
+
+        return self._get_cycle_param()
+
+    def _get_cycle_param(self):
+        raise NotImplementedError
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:367]: We could add the docstring here as well:
```python
"""Method to get the cycle's current parameter value"""
```

**Review by sadra-barikbin — COMMENTED** [review]
Thank you @sihyeong671 , I left some comments.

File: ignite/handlers/param_scheduler.py [hunk 8]
```diff
@@ -325,18 +339,33 @@ def __init__(
...
         return super(CyclicalScheduler, self).__call__(engine, name)
 
+    def get_param(self) -> float:
+        """Method to get current optimizer's parameter value"""
+        if self.warmup_each_cycle and self.event_index < self.warmup_duration:
+            return self.end_value + (self.start_value - self.end_value) * self.event_index / self.warmup_duration
+
+        return self._get_cycle_param()
+
+    def _get_cycle_param(self):
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:366]:
```suggestion
    def _get_cycle_param(self) -> float:
```

File: ignite/handlers/param_scheduler.py [hunk 9]
```diff
@@ -313,6 +315,18 @@ def __init__(
...
+        if not self.warmup_each_cycle:
+            if warmup_duration is not None:
+                warnings.warn(
+                    f"warmup_each_cycle=False but your warmup_duration is {warmup_duration}. "
+                    "so warmup_duration will be set to 0. "
+                    "If you want to use warmup each cycle, pleas set warmup_each_cycle=True"
+                )
+            self.warmup_duration = 0
+        else:
+            self.warmup_duration = warmup_duration
```
Comment (sadra-barikbin) [inline ignite/handlers/param_scheduler.py:328]: What if `warmup_duration` is `None`? In the next line we add it to `self.cycle_size`

File: ignite/handlers/param_scheduler.py (hunk 8 above)
Comment (vfdev-5) [reply ignite/handlers/param_scheduler.py:366]: @sihyeong671 To implement an abstract class please use this pattern: https://github.com/pytorch/ignite/blob/ca3033a7ac30d6afae4eb6f5a03d6a6ffad49c84/ignite/handlers/param_scheduler.py#L89-L96

Comment (sihyeong671) [reply ignite/handlers/param_scheduler.py:366]: > @sihyeong671 To implement an abstract class please use this pattern:
> 
> https://github.com/pytorch/ignite/blob/ca3033a7ac30d6afae4eb6f5a03d6a6ffad49c84/ignite/handlers/param_scheduler.py#L89-L96

@vfdev-5 @sadra-barikbin 
Thank for review my code!
you mean change below code to above pattern on line89?
```py
def _get_cycle_param(self) -> float:
        """Method to get the cycle's current parameter value"""
        raise NotImplementedError
```

File: ignite/handlers/param_scheduler.py (hunk 9 above)
Comment (sihyeong671) [reply ignite/handlers/param_scheduler.py:328]: > What if `warmup_duration` is `None`? In the next line we add it to `self.cycle_size`

oh, I didn't care about that. I will add code that handle when warmup_duration is none!

**Comment by vfdev-5** [comment]
@sihyeong671 thanks for working on this PR! Can you please fix this test issue:
```
FAILED tests/ignite/handlers/test_param_scheduler.py::test_cyclical_scheduler_asserts - AssertionError: Regex pattern did not match.
 Regex: "Can't instantiate abstract class CyclicalScheduler with abstract method _get_cycle_param"
 Input: "Can't instantiate abstract class CyclicalScheduler with abstract methods _get_cycle_param"
```

File: ignite/handlers/param_scheduler.py [hunk 10]
```diff
@@ -287,7 +287,8 @@ class CyclicalScheduler(ParamScheduler):
         If the scheduler is bound to an 'ITERATION_*' event, 'cycle_size' should
         usually be the number of batches in an epoch.
 
-    .. versionadded:: 0.4.5
+    .. versionchanged:: 0.4.13
+        Added warmup to the scheduler using ``warmup_each_cycle`` and ``warmup_duration``.
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:291]: We should keep previous tag:
```suggestion
.. versionadded:: 0.4.5

.. versionchanged:: 0.4.13
        Added warm-up option to the scheduler using ``warmup_each_cycle`` and ``warmup_duration``.
```

File: ignite/handlers/param_scheduler.py [hunk 11]
```diff
@@ -432,7 +470,8 @@ def print_lr():
     .. versionadded:: 0.4.5
     """
 
-    def get_param(self) -> float:
+    def _get_cycle_param(self) -> float:
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:473]: please rename `get_param -> _get_cycle_param` in `CosineAnnealingScheduler` class as well.

**Comment by sihyeong671** [comment]
> @sihyeong671 thanks for working on this PR! Can you please fix this test issue:
> 
> ```
> FAILED tests/ignite/handlers/test_param_scheduler.py::test_cyclical_scheduler_asserts - AssertionError: Regex pattern did not match.
>  Regex: "Can't instantiate abstract class CyclicalScheduler with abstract method _get_cycle_param"
>  Input: "Can't instantiate abstract class CyclicalScheduler with abstract methods _get_cycle_param"
> ```

In macOS, If I change the sentence, AssertionError like below code
First, I will change it and PR
```
E           AssertionError: Regex pattern did not match.
E            Regex: "Can't instantiate abstract class CyclicalScheduler with abstract methods _get_cycle_param"
E            Input: "Can't instantiate abstract class CyclicalScheduler with abstract method _get_cycle_param"
```

**Comment by vfdev-5** [comment]
@sihyeong671 I see, let's just check for this sting "Can't instantiate abstract class CyclicalScheduler":
```python
    with pytest.raises(
        TypeError, match="Can't instantiate abstract class CyclicalScheduler"
    ):
        CyclicalScheduler({}, "lr", 0.0, 1.0, 10)
```

**Comment by sihyeong671** [comment]
@vfdev-5 I have no idea how to fix the error. error shows "Error: The operation was canceled.", how to handle it?

**Comment by vfdev-5** [comment]
@sihyeong671 no worries, this can be a flaky CI. We set 45 minutes limit if the test stuck with distributed tests. So, just restarting it.

File: ignite/handlers/param_scheduler.py [hunk 12]
```diff
@@ -313,6 +318,20 @@ def __init__(
         self.cycle = 0
         self.start_value_mult = start_value_mult
         self.end_value_mult = end_value_mult
+        self.warmup_each_cycle = warmup_each_cycle
+        if not self.warmup_each_cycle:
+            if warmup_duration is not None:
+                warnings.warn(
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:324]: @sihyeong671 actually why we should warn here? For me these two args `warmup_duration` and `warmup_each_cycle` are less tightly coupled:
```
warmup_duration is None
warmup_each_cycle can be True or False
=> no warm-up at all and nothing to warn about

warmup_duration is not None
warmup_each_cycle can be True or False
=> add a warm-up
```
What do you think ?

Comment (sihyeong671) [reply ignite/handlers/param_scheduler.py:324]: Now I think `warmup_each_cycle` variable is not necessary. I will remove `warmup_each_cycle`!

File: ignite/handlers/param_scheduler.py [hunk 13]
```diff
@@ -538,7 +567,7 @@ def print_lr():
 
     def get_param(self) -> float:
         """Method to get current optimizer's parameter value"""
-        cycle_progress = self.event_index / self.cycle_size
+        cycle_progress = (self.event_index - self.warmup_duration) / self.cycle_size
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:570]: should not we apply the same logic to `LinearCyclicalScheduler.get_param` ?

**Comment by vfdev-5** [comment]
@sihyeong671 can you please fix mypy issue: https://github.com/pytorch/ignite/actions/runs/6231400451/job/16912898300?pr=3064#step:12:18

File: ignite/handlers/param_scheduler.py [hunk 14]
```diff
@@ -308,11 +320,13 @@ def __init__(
         )
         self.start_value = start_value
         self.end_value = end_value
-        self.cycle_size = int(cycle_size)  # Ensure cycle_size is integer
+        self.cycle_size = cycle_size
         self.cycle_mult = cycle_mult
         self.cycle = 0
         self.start_value_mult = start_value_mult
         self.end_value_mult = end_value_mult
+        self.warmup_duration = cyclic_warmup_duration
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:328]: Nit: why you have to names for this arg: warmup_duration and cyclic_warmup_duration ? Can we simply use warmup_duration everywhere?

File: ignite/handlers/param_scheduler.py [hunk 15]
```diff
@@ -288,6 +296,9 @@ class CyclicalScheduler(ParamScheduler):
         usually be the number of batches in an epoch.
 
     .. versionadded:: 0.4.5
+
+    .. versionchanged:: 0.4.13
+        Added cyclic warm-up to the scheduler using ``warmup_each_cycle`` and ``warmup_duration``.
```
Comment (vfdev-5) [inline ignite/handlers/param_scheduler.py:301]: This note is outdated

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @sihyeong671 and @sadra-barikbin !

**Review by sadra-barikbin — APPROVED** [review]

**Pull request merged**